metagraph/src/graph/graph_extensions/node_rc.cpp

```diff
@@ -0,0 +1,304 @@
+#include "node_rc.hpp"
+
+#include "graph/representation/succinct/dbg_succinct.hpp"
+#include "common/seq_tools/reverse_complement.hpp"
+#include "common/vectors/bitmap_builder.hpp"
+#include "common/utils/template_utils.hpp"
+
+#include <progress_bar.hpp>
+#include <ips4o.hpp>
+#include <mutex>
+
+namespace mtg {
+namespace graph {
+
+using boss::BOSS;
+using edge_index = BOSS::edge_index;
+
+template <class Indicator, class Mapping>
+NodeRC<Indicator, Mapping>::NodeRC(const DBGSuccinct &graph)
+      : graph_(&graph) {
+    if (graph.get_mode() != DeBruijnGraph::PRIMARY) {
+        common::logger->error("Only implemented for PRIMARY graphs");
+        exit(1);
+    }
+
+    const BOSS &boss = graph.get_boss();
+
+    std::unique_ptr<bitmap_builder> builder;
+    builder = std::make_unique<bitmap_builder_set>(graph.max_index() + 1, 0,
+                                                   graph.num_nodes() * 0.06);
+
+    std::mutex mu;
+
+    std::vector<std::pair<node_index, std::pair<edge_index, edge_index>>> mapping;
+    graph.call_sequences([&](const std::string &seq, const std::vector<node_index> &path) {
+        std::string rev_seq = seq;
+        ::reverse_complement(rev_seq.begin(), rev_seq.end());
+        auto encoded = boss.encode(rev_seq);
+
+        // initialize
+        auto it = encoded.end() - boss.get_k();
+        auto [edge, edge_2, end] = boss.index_range(it, it + boss.get_k());
+        assert(end != it + boss.get_k() || edge == edge_2);
+
+        for (size_t i = 0; i < path.size(); ++i) {
+            // prefix
+            size_t map_i = 0;
+            bool found = false;
+            if (end == it + boss.get_k()) {
+                std::lock_guard<std::mutex> lock(mu);
+                assert(edge);
+                map_i = mapping.size();
+                mapping.emplace_back(path[i], std::make_pair(edge, 0));
+                found = true;
+                builder->add_one(path[i]);
+            }
+
+            // suffix
+            --it;
+            std::tie(edge, edge_2, end) = boss.index_range(it, it + boss.get_k());
+            assert(end != it + boss.get_k() || edge == edge_2);
+
+            if (end == it + boss.get_k()) {
+                std::lock_guard<std::mutex> lock(mu);
+                assert(edge);
+                if (found) {
+                    assert(mapping[map_i].first == path[i]);
+                    assert(mapping[map_i].second.first);
+                    assert(!mapping[map_i].second.second);
+                    mapping[map_i].second.second = edge;
+                } else {
+                    builder->add_one(path[i]);
+                    mapping.emplace_back(path[i], std::make_pair(0, edge));
+                }
+            }
```

Comment:
add more comments

Reply:
I cleaned this up a lot, and added comments